### PR TITLE
Added IAP scene and IAP functionality

### PR DESCRIPTION
**Pull request by timdng10**
Added IAP scene and IAP functionality

### CODE REVIEW

File: Assets/Scripts/AdsAndIAP/IAPController.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        get
+        {
+            return Instance as IAPController;
+        }
+    }
+
+    #endregion
+
+    #region MonoBehaviourExtended Overrides
+    protected override void fetchReferences()
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/IAPController.cs]: Newline after #region

File: Assets/Scripts/AdsAndIAP/IAPController.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        base.fetchReferences();
+        InitializePurchasing();
+    }
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/IAPController.cs]: Newline before #endregion

File: Assets/Scripts/AdsAndIAP/IAPController.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    #endregion
+
+    #region IStoreListener Overrides
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/IAPController.cs]: Newline after #region

Also the convention when it's an interface is to write "IStoreListener Interface"

File: Assets/Scripts/AdsAndIAP/IAPController.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        return PurchaseProcessingResult.Complete;
+    }
+
+    public void OnPurchaseFailed(Product product, PurchaseFailureReason failureReason)
+    {
+        Debug.LogError(string.Format("OnPurchaseFailed: FAIL. Product: '{0}', PurchaseFailureReason: {1}", product.definition.storeSpecificId, failureReason));
+    }
+    #endregion
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/IAPController.cs]: Add newline before #endregion

File: Assets/Scripts/AdsAndIAP/IAPController.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    private bool IsInitialized()
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/IAPController.cs]: We generally leave off the `private` access modifier as it's implicit. Also private functions are camelCase "isInitialized"

File: Assets/Scripts/AdsAndIAP/IAPController.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            return;
+        }
+
+        var builder = ConfigurationBuilder.Instance(StandardPurchasingModule.Instance());
+        builder.AddProduct(PRODUCT_1000_GOLD, ProductType.Consumable);
+
+        UnityPurchasing.Initialize(this, builder);
+    }
+
+    void BuyProductID(string productId)
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/IAPController.cs]: Should be camelCase

File: Assets/Scripts/AdsAndIAP/IAPController.cs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                Debug.LogError("BuyProductID: FAIL. Not purchasing product, either is not found or is not available for purchase");
+            }
+        }
+        else
+        {
+            Debug.LogError("BuyProductID FAIL. Not initialized.");
+        }
+    }
+
+    public void PurchaseGold1000()
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/IAPController.cs]: Should use the term "coins" to be consistent with the rest of the codebase

File: Assets/Scripts/AdsAndIAP/IAPController.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+ */
+
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Purchasing;
+
+public class IAPController : SingletonController<IAPController>, IStoreListener
+{
+    const string PRODUCT_1000_GOLD = "gold1000";
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/IAPController.cs]: Should use the term "coins" to be consistent with the rest of the codebase

**Comment by timdng10** [comment]
The code cant run on the device until the apps are put up in the appstore/playstore 
Then more things have to be set up through their own system, so google and apple know how much money we owe them

**Comment by imann24** [comment]
@timdng10 The store setup is potentially part of this task. I'll run this by Grace and Sam.

File: Assets/Scripts/AdsAndIAP/IAPController.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    public void OnPurchaseFailed(Product product, PurchaseFailureReason failureReason)
+    {
+        Debug.LogError(string.Format("OnPurchaseFailed: FAIL. Product: '{0}', PurchaseFailureReason: {1}", product.definition.storeSpecificId, failureReason));
+    }
+
+    #endregion
+
+    bool IsInitialized()
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/IAPController.cs]: Should be camelCase

**Pull request merged**